src/librustc_mir/transform/promote_consts.rs

```diff
@@ -748,7 +748,14 @@ impl<'tcx> Validator<'_, 'tcx> {
     ) -> Result<(), Unpromotable> {
         let fn_ty = callee.ty(self.body, self.tcx);
 
-        if !self.explicit && self.const_kind.is_none() {
+        // `const` and `static` use the explicit rules for promotion, meaning calls to `const fn`
+        // can be promoted.
+        let context_uses_explicit_promotion_rules = matches!(
+            self.const_kind,
+            Some(hir::ConstContext::Static(_) | hir::ConstContext::Const)
+        );
+
+        if !self.explicit && !context_uses_explicit_promotion_rules {
```

Comment:
`self.explicit` is already encoding whether we use explicit promotion, and that's what we check everywhere else in this file. Why does this one check need special treatment?

Reply:
Ah I see, it's the `self.const_kind.is_none()` here that introduced the exception. IMO that should just be `if !self.explicit`... but that doesn't seem how we currently organize things.

Reply:
However, this means that all `const_kind.is_none()` are wrong. Promotion should behave the same for `const_kind == Some(ConstFn)` and `const_kind == None`.

Reply:
Most of these checks are for things that are not allowed in a const context (without feature gates). Presumably this logic is left over from when promotion and const-checking happened in the same module.

Reply:
Ah, that's a good point.

Still, it wouldn't hurt to adjust them all to call something like `is_const_only` or so (meaning "this code will only ever be executed during CTFE), just to make sure we haven't missed any of them.